src/server/pfs/server/source.go

```diff
@@ -0,0 +1,255 @@
+	h := pfs.NewHash()
+	for _, dataRef := range idx.DataOp.DataRefs {
+		// TODO Pull from chunk hash.
+		h.Write([]byte(dataRef.Hash))
+	}
+	return &FileReader{
+		file: file,
+		idx:  idx,
+		fmr:  fmr,
+		mr:   mr,
+		hash: h,
+	}
+}
+
+func (fr *FileReader) updateFileInfo(idx *index.Index) {
+	fr.fileCount++
+	for _, dataRef := range idx.DataOp.DataRefs {
+		fr.hash.Write([]byte(dataRef.Hash))
+	}
+}
+
+// Info returns the info for the file.
+func (fr *FileReader) Info() *pfs.FileInfoV2 {
+	return &pfs.FileInfoV2{
+		File: fr.file,
+		Hash: pfs.EncodeHash(fr.hash.Sum(nil)),
+	}
+}
+
+// Get writes a tar stream that contains the file.
+func (fr *FileReader) Get(w io.Writer, noPadding ...bool) error {
+	if err := fr.fmr.Get(w); err != nil {
+		return err
+	}
+	for fr.fileCount > 0 {
+		fmr, err := fr.mr.Next()
+		if err != nil {
+			return err
+		}
+		if err := fmr.Get(w); err != nil {
+			return err
+		}
+		fr.fileCount--
+	}
+	if len(noPadding) > 0 && noPadding[0] {
+		return nil
+	}
+	// Close a tar writer to create tar EOF padding.
+	return tar.NewWriter(w).Close()
+}
+
+func (fr *FileReader) drain() error {
+	for fr.fileCount > 0 {
+		if _, err := fr.mr.Next(); err != nil {
+			return err
+		}
+		fr.fileCount--
+	}
+	return nil
+}
+
+// Source iterates over FileInfoV2s generated from a fileset.Source
+type Source struct {
+	commit        *pfs.Commit
+	getReader     func() fileset.FileSource
+	computeHashes bool
+}
+
+// Creates a Reader which emits FileInfoV2s with the information from commit, and the entries from readers
+// returned by getReader.  If getReader returns different Readers all bets are off.
+func NewSource(commit *pfs.Commit, computeHashes bool, getReader func() fileset.FileSource) *Source {
+	return &Source{
+		commit:        commit,
+		getReader:     getReader,
+		computeHashes: computeHashes,
+	}
+}
+
+func (s *Source) Iterate(ctx context.Context, cb func(*pfs.FileInfoV2, fileset.File) error) error {
+	ctx, cf := context.WithCancel(ctx)
+	defer cf()
+	fs1 := s.getReader()
+	fs2 := s.getReader()
+	s2 := newStream(ctx, fs2)
+	cache := make(map[string][]byte)
+	return fs1.Iterate(ctx, func(fr fileset.File) error {
+		idx := fr.Index()
+		fi := &pfs.FileInfoV2{
+			File: client.NewFile(s.commit.Repo.Name, s.commit.ID, idx.Path),
+		}
+		if s.computeHashes {
+			var err error
+			var hashBytes []byte
+			if indexIsDir(idx) {
+				hashBytes, err = computeHash(cache, s2, idx.Path)
+				if err != nil {
+					return err
+				}
+			} else {
+				hashBytes = computeFileHash(idx)
+			}
+			fi.Hash = pfs.EncodeHash(hashBytes)
+		}
+		if err := cb(fi, fr); err != nil {
+			return err
+		}
+		delete(cache, idx.Path)
+		return nil
+	})
+}
+
+func computeHash(cache map[string][]byte, s *stream, x string) ([]byte, error) {
+	if hashBytes, exists := cache[x]; exists {
+		return hashBytes, nil
+	}
+	// check that the next item in the stream is x
```

Comment:
I don't think we need to peek. I think we can just progress the stream then error if it is not in the correct place. I don't think having the streams out of sync will be recoverable.